agent/dockerclient/dockerauth/asm.go

```diff
@@ -0,0 +1,171 @@
+	registryAuthData *api.RegistryAuthenticationData) (docker.AuthConfiguration, error) {
+
+	if registryAuthData == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("dockerauth: missing container registry's auth data")
+	}
+
+	authData := registryAuthData.ASMAuthData
+
+	if authData == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("dockerauth: missing container registry's aws secrets manager auth data")
+	}
+
+	// First try to get the secret from cache, if the secret does not exist,
+	// then call ASM api to get the new secret
+	key := asmCacheKey{
+		taskARN:              authData.GetTaskARN(),
+		credentialsParameter: authData.CredentialsParameter,
+	}
+
+	// Try to get the auth config from cache
+	auth := authProvider.getAuthConfigFromCache(key)
+	if auth != nil {
+		return *auth, nil
+	}
+
+	// Get the auth config from ASM
+	return authProvider.getAuthConfigFromASM(image, key, authData)
+}
+
+// getAuthconfigFromCache retrieves the docker.AuthConfiguration from cache
+func (authProvider *asmAuthProvider) getAuthConfigFromCache(key asmCacheKey) *docker.AuthConfiguration {
+
+	authData, ok := authProvider.cache.Get(key.String())
+	if !ok {
+		return nil
+	}
+
+	cachedAuthData, ok := authData.(*docker.AuthConfiguration)
+	if !ok {
+		log.Warnf("Reading ASM credentials from cache failed")
+		return nil
+	}
+
+	return cachedAuthData
+}
+
+// getAuthConfigFromASM calls the ASM API to get docker auth config
+func (authProvider *asmAuthProvider) getAuthConfigFromASM(image string,
+	key asmCacheKey, authData *api.ASMAuthData) (docker.AuthConfiguration, error) {
+
+	in := &secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(authData.CredentialsParameter),
+	}
+
+	log.Debugf("Calling ASM.GetSecretValue for %s", image)
+	out, err := authProvider.client.GetSecretValue(in)
+	if err != nil {
+		return docker.AuthConfiguration{}, err
+	}
+
+	if out == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("asm auth: missing AuthorizationData in ASM response for %s", image)
+	}
+
+	dac, err := authProvider.extractASMValue(out)
+	if err != nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("asm auth: failed to extract AuthorizationData from ASM response for %s, %v", image, err)
+	}
+
+	// Cache the new token
+	authProvider.cache.Set(key.String(), &dac)
+	return dac, nil
+}
+
+func (authProvider *asmAuthProvider) extractASMValue(out *secretsmanager.GetSecretValueOutput) (docker.AuthConfiguration, error) {
+	if out.SecretString == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("asm auth: ASM response missing SecretString")
+	}
+
+	authDataValue := &asmAuthDataValue{}
+
+	secretstring := aws.StringValue(out.SecretString)
+	err := json.Unmarshal([]byte(secretstring), authDataValue)
+	if err != nil {
+		// could not unmarshal, incorrect secret value schema
+		return docker.AuthConfiguration{}, fmt.Errorf("asm auth: AuthorizationData is malformed")
+	}
+
+	if authDataValue.Username == nil {
```

Comment:
Is it valid for either of `Username`/`Password` fields to be empty strings?

Reply:
i should check, i suspect it would just return wrong username/password. 

are you asking if we should error right here if empty strings for these fields are known to be invalid?

Reply:
> are you asking if we should error right here if empty strings for these fields are known to be invalid?

Yeah, that.